src/roinput.rs

```diff
@@ -0,0 +1,856 @@
+use super::*;
+use ark_ff::PrimeField;
+use bitvec::{prelude::*, view::AsBits};
+
+pub trait Input: Copy {
+    fn to_roinput(self) -> ROInput;
+    // The domain string length must be <= 20 (TODO: Limit at compile time)
+    fn domain_string(self, network_id: NetworkId) -> &'static str;
+}
+
+pub struct ROInput {
+    pub fields: Vec<PallasField>,
+    pub bits: BitVec<Lsb0, u8>,
+}
+
+impl ROInput {
+    pub fn new() -> Self {
+        ROInput {
+            fields: vec![],
+            bits: BitVec::new(),
+        }
+    }
+
+    pub fn append_field(&mut self, f: PallasField) {
+        self.fields.push(f);
+    }
+
+    pub fn append_scalar(&mut self, s: PallasScalar) {
+        // mina scalars are 255 bytes
+        let bytes = s.to_bytes(); // TODO: Combine these two into one-liner
+        let bits = &bytes.as_bits::<Lsb0>()[..PallasScalar::size_in_bits()];
+        self.bits.extend(bits);
+    }
+
+    pub fn append_bit(&mut self, b: bool) {
+        self.bits.push(b);
+    }
+
+    pub fn append_bytes(&mut self, bytes: Vec<u8>) {
+        self.bits.extend_from_bitslice(bytes.as_bits::<Lsb0>());
+    }
+
+    pub fn append_u32(&mut self, x: u32) {
+        self.append_bytes(x.to_le_bytes().to_vec());
+    }
+
+    pub fn append_u64(&mut self, x: u64) {
+        self.append_bytes(x.to_le_bytes().to_vec());
+    }
+
+    pub fn to_bytes(&mut self) -> Vec<u8> {
+        let mut bits: BitVec<Lsb0, u8> = self.fields.iter().fold(BitVec::new(), |mut acc, fe| {
+            acc.extend_from_bitslice(
+                &fe.to_bytes().as_bits::<Lsb0>()[..PallasField::size_in_bits()],
```

Comment:
That's the mina signature spec.  We only take the bits that are actually used to represent the field element, i.e. we pack the bits instead of just being byte aligned.